src/EffectInfo.cpp

```diff
@@ -88,6 +93,62 @@ EffectBase* EffectInfo::CreateEffect(string effect_type) {
 	return NULL;
 }
 
+#if defined(__linux__)
+EffectBase* EffectInfo::LoadEffect(std::string location){
+    pthread_mutex_lock(&m_mutex);
+
+    void * file = dlopen(location.c_str(), RTLD_NOW);
+    if (file == nullptr){
+        pthread_mutex_unlock(&m_mutex);
+        throw InvalidFile("Can not open file", location.c_str());
+    }
+    void * factory_ptr = dlsym(file, "factory");
+
+    if (factory_ptr == nullptr){
+        pthread_mutex_unlock(&m_mutex);
+        dlclose(file);
+        throw InvalidFile("Can not find requested plugin API in file", location.c_str());
+    }
+
+    EffectBase * (*factory)(uint16_t);
+    factory = (EffectBase* (*)(uint16_t)) factory_ptr;
+
+    EffectBase* instance = factory(OPENSHOT_PLUGIN_API_VERSION);
+
+    if (instance == nullptr){
+        pthread_mutex_unlock(&m_mutex);
+        dlclose(file);
+        throw InvalidFile("Plugin does not support current version of openshot", location.c_str());
+    }
+
+    m_loadedDynamicEffects.insert(make_pair(std::string(instance->info.name), factory));
+    m_loadedDynamicHandles.insert(m_loadedDynamicHandles.end(), file);
+
+    pthread_mutex_unlock(&m_mutex);
+    return instance;
+}
+
+void EffectInfo::UnloadDynamicEffects(){
+    pthread_mutex_lock(&m_mutex);
+
+    for (auto & handle : m_loadedDynamicHandles){
+        dlclose(handle);
+    }
+
+    m_loadedDynamicEffects.clear();
+    m_loadedDynamicHandles.clear();
+    pthread_mutex_unlock(&m_mutex);
+}
+#else
+EffectBase* EffectInfo::LoadEffect(std::string location){
+    return NULL;
+}
+
+void EffectInfo::UnloadDynamicEffects(){
+
+}
+#endif
+
```

Comment:
This might be cleaner/easier to write as...

```cpp
EffectBase* EffectInfo::LoadEffect(std::string location){
#if !defined(__linux__)
    return NULL;
#endif
    pthread_mutex_lock(&m_mutex);
    ...
}

void EffectInfo::UnloadDynamicEffects(){
#if defined(__linux__)
    pthread_mutex_lock(&m_mutex);

    for (auto & handle : m_loadedDynamicHandles){
        dlclose(handle);
    }

    m_loadedDynamicEffects.clear();
    m_loadedDynamicHandles.clear();
    pthread_mutex_unlock(&m_mutex);
#endif
}
```

Primary advantage: If the signature of `EffectInfo::LoadEffect()` or `EffectInfo::UnloadDynamicEffects()` changes, we don't have to worry about updating it in two places and making sure they're kept in sync.

Reply:
(Sorry, accidentally had an earlier version of this comment included in the review, along with the final one. I've deleted the redundant comment, but the email notification version may have been a bit confusing.)